lib/core/test/integration/Test/Integration/Framework/DSL.hs

```diff
@@ -1010,6 +1015,20 @@ postTransactionFeeViaCLI ctx args = do
         err <- TIO.hGetContents stderr
         return (c, T.unpack out, err)
 
+listTransactionsViaCLI
+    :: forall t r s . (CmdResult r, HasType Port s, KnownCommand t)
+    => s
+    -> ApiWallet
+    -> Maybe Iso8601Time
+    -> Maybe Iso8601Time
```

Comment:
I think it would be better to have just a _[String]_ as an argument of this method. This would allow for more flexibility when constructing tests, in particular we can add a lot of different negative tests (e.g. faulty wallet ID, _typos_ in the parameter names, bad-formed parameter values etc.) which is not possible currently.

Reply:
> I think it would be better to have just a _[String]_ as an argument of this method. This would allow for more flexibility when constructing tests, in particular we can add a lot of different negative tests (e.g. faulty wallet ID, _typos_ in the parameter names, bad-formed parameter values etc.) which is not possible currently.

@piotr-iohk thanks again for looking at this.

I can sympathize with this argument. Though for this PR, I consciously _avoided_ doing that due to previous feedback: for example: https://github.com/input-output-hk/cardano-wallet/pull/519#discussion_r302404803

@KtorZ, would you be able to share your opinion on this?

Regarding the use of negative tests, I had some thoughts which I wanted to share here:

**Regarding simple typos in parameter names**: rather than writing lots of manual tests (which are likely to be rather repetitive), perhaps there is a better, automated way to generate CLI calls with mis-spelt parameter names? We could potentially generate these from the CLI definition.

**Regarding badly-formed parameters that the CLI can detect due to parsing errors** (i.e., `fromText` returning `Left`). Again, perhaps there is a way to generate these automatically? (Using QuickCheck, we can generate arbitrary values that we know will not parse with `fromText`, and pass these to the CLI, expecting failure.)

**Regarding invalid parameters that parse correctly, whose invalidity can only be detected by the back-end**. Because the CLI must first parse data from `Text` to well-typed values before sending it to the back end, I suspect this case also doesn't require the use of `String` types.

@piotr-iohk Would be interested to hear your opinion!

Reply:
Thanks @jonathanknowles!
 
I think the way I'm coming from is just the past experience: In https://github.com/input-output-hk/cardano-wallet-legacy we actually had a strongly typed API client and it used to be quite limiting for me in terms of flexibility to create test cases (in particularly negative) (see examplary request -> https://github.com/input-output-hk/cardano-wallet-legacy/blob/develop/test/integration/Test/Integration/Scenario/Transactions.hs#L255-L262). As an aid there was this _unsafeRequest_ added later which was less _typed_ but it was giving more flexibility (e.g. https://github.com/input-output-hk/cardano-wallet-legacy/blob/develop/test/integration/Test/Integration/Scenario/EosWallets.hs#L59-L69). So in this iteration of wallet we kind of went for the second approach from the beggining. 

In my mind the idea of automated integration test is just to have a tool to be able to make the same calls and get the same responses (and test them) as the end user would (and user can basically type anything ;))
Off course if there is a room for improving this, I'm all for it (in particular for generating all sort of negative test cases). The only thing that I would not want to achieve is that any additional layers of abstraction are not in fact something that limit us in testing.

Reply:
Having a stringly-typed API makes for a quite poor documentation when reading function signatures. Yet, having said that, there's not much we do with the underlying data-type apart from "showing it". Plus, we do want to submit values that are not valid and therefore, that couldn't be represented as an `Iso8601Range`. Most of the integration tests work this way for this very reason as @piotr-iohk highlighted already.

Regarding the automated generation of false cases at the integration level: sounds like _overkill_ and probably too complicated at this stage. We can already test all sort of things at the unit level with extensive properties and all sort of negative tests. Plus, integration tests are already quite complicated on their own, so we have to strive to avoid this complexity from exploding. Mixing property tests at this stage would require some careful thoughts and design for something we can already do very well at the unit level.

Reply:
@KtorZ @piotr-iohk 

Okay. I've:

- [x] changed the integration test to use a matrix of hand-picked time values (rather than using a property-based test).
- [x] changed the `listTransactionsViaCLI` function to take `Maybe String` for each time value.

Hopefully this should address the concerns raised above, but let me know if you'd like me to change anything about this.

Reply:
@KtorZ  wrote:

> Regarding the automated generation of false cases at the integration level: sounds like _overkill_ and probably too complicated at this stage.

My point is that "typos in the parameter names" was used as a justification for allowing parameters to be supplied as `[String]` in the `listTransactionsViaCLI` command:

Assuming we decide that it's valuable to test that parameter name _misspellings_ are rejected, wouldn't it be better to _generate_ such tests automatically from the CLI definition? (as part of a unit test suite.)

It seems like there would be countless permutations, especially when we take all commands and argument combinations into account.

Reply:
Btw. another advantage of `[String]` is also that we can test different order of parameters (--start ... --end ... or --end ... --start ...)